From baa66de2bc6aa8d465d410230ddcef940cdd8396 Mon Sep 17 00:00:00 2001
From: Shadi Ammouri <shadi@marvell.com>
Date: Mon, 9 Oct 2017 13:55:55 +0300
Subject: [PATCH 2021/2241] dma: mv_xor_v2: Add DT property to skip DMA
 registeration for user-space driver usage

Add "musdk-mode" DT property that will cause device probing to fail, and
thus prevent contention between the in-kernel DMA driver and the
respective user-space driver for the same DMA engine.

Change-Id: I2ce401f8840195829254e260f9cd3711d44c3921
Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/45074
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 .../devicetree/bindings/dma/mv-xor-v2.txt          | 26 +++++++++++++++++++++-
 drivers/dma/mv_xor_v2.c                            |  5 +++++
 2 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/Documentation/devicetree/bindings/dma/mv-xor-v2.txt b/Documentation/devicetree/bindings/dma/mv-xor-v2.txt
index 006699e..772e425 100644
--- a/Documentation/devicetree/bindings/dma/mv-xor-v2.txt
+++ b/Documentation/devicetree/bindings/dma/mv-xor-v2.txt
@@ -8,9 +8,19 @@ Required properties:
 - dma-bus-width: DMA address bus width
 - msi-parent: Phandle to the MSI-capable interrupt controller used for
   interrupts.
-- clocks: One ore more clock gating clocks to be used to enable DMA engine
+- clocks: One or more clock gating clocks to be used to enable DMA engine
   unit in the SoC.
 
+Optional properties:
+The below is relevant only if Marvell Userspace SDK (MUSDK) is in use:
+- musdk-mode (bool): Indicates whether the device is to be controlled by musdk framework.
+  In this case, the probe process of the device is skipped, in order not to create
+  a contention with the user-space drivers.
+  In some cases, when musdk mode is enabled, the device will be attached to a generic
+  uio instance, and this is why its not marked as "status=disabled".
+  For more info about attaching the device to generic-uio instance, see
+  Documentation/devicetree/bindings/uio/uio_pdrv_genirq.txt
+
 Example:
 
 	xor0@400000 {
@@ -22,3 +32,17 @@ Example:
 		dma-coherent;
 		dma-width = <40>;
 	};
+
+
+	/* musdk-mode enabled example */
+	xor0@400000 {
+		compatible = "marvell,mv-xor-v2";
+		reg = <0x400000 0x1000>,
+		      <0x410000 0x1000>;
+		msi-parent = <&gic_odmi>;
+		clocks = <&gateclk 14>;
+		dma-coherent;
+		musdk-mode;
+		uio-name = "uio_dma_xor_0";
+		dma-width = <40>;
+	};
diff --git a/drivers/dma/mv_xor_v2.c b/drivers/dma/mv_xor_v2.c
index d81d5b3..3406f9f 100644
--- a/drivers/dma/mv_xor_v2.c
+++ b/drivers/dma/mv_xor_v2.c
@@ -891,6 +891,11 @@ static int mv_xor_v2_probe(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	u32 dma_bus_width;
 
+	if (of_property_read_bool(pdev->dev.of_node, "musdk-mode")) {
+		dev_notice(&pdev->dev, "Marvell Version 2 XOR driver - MUSDK mode\n");
+		return -1;
+	}
+
 	dev_notice(&pdev->dev, "Marvell Version 2 XOR driver\n");
 
 	xor_dev = devm_kzalloc(&pdev->dev, sizeof(*xor_dev), GFP_KERNEL);
-- 
2.7.4

